File: 47-permutations-ii/47-permutations-ii.py

def findPermutation(res,nums,numsLength,pos):

    if(pos==numsLength):
        res.append(nums[:])
    else:
        posMap = set()
        for idx in range(pos,numsLength):
            if nums[idx] in posMap: continue
            posMap.add(nums[idx])
            nums[idx], nums[pos] = nums[pos],nums[idx]
            findPermutation(res,nums,numsLength,pos+1)
            nums[idx], nums[pos] = nums[pos],nums[idx]


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
        numsLength = len(nums)
        pos = 0
        res = []
        findPermutation(res,nums,numsLength,pos)
        return res

File: 47-permutations-ii/47-permutations-ii.cpp
class Solution {
public:
    void findSolutions(vector<int> nums, int currIndex,vector<vector<int>> &solutions ){

        if(currIndex==nums.size()){
            solutions.push_back(nums);
            return;
        }
        
        int n = nums.size();
        unordered_set<int> hashSet;
        for(int i=currIndex;i<n;i++){
            if(hashSet.find(nums[i])==hashSet.end()){
                hashSet.insert(nums[i]);
                swap(nums[currIndex],nums[i]);
                findSolutions(nums,currIndex+1,solutions);
                swap(nums[currIndex],nums[i]);
            }
        }
        
    }
    
    
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        
        vector<vector<int>> solutions;
        int currIndex = 0;
        findSolutions(nums,currIndex,solutions);
        return solutions;
        
    }
};